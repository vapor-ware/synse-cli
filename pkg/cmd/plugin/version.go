// Synse CLI
// Copyright (c) 2019 Vapor IO
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

package plugin

import (
	"context"
	"io"

	"github.com/spf13/cobra"
	"github.com/vapor-ware/synse-cli/pkg/utils"
	"github.com/vapor-ware/synse-cli/pkg/utils/exit"
	synse "github.com/vapor-ware/synse-server-grpc/go"
)

func init() {
	cmdVersion.Flags().BoolVarP(&flagNoHeader, "no-header", "n", false, "do not print out column headers")
	cmdVersion.Flags().BoolVarP(&flagJSON, "json", "", false, "print output as JSON")
	cmdVersion.Flags().BoolVarP(&flagYaml, "yaml", "", false, "print output as YAML")
}

var cmdVersion = &cobra.Command{
	Use:   "version",
	Short: "Display version information for the plugin",
	Long: utils.Doc(`
		Display version information for the plugin.

		The output of this command can be formatted as a table (default), as
		JSON, or as YAML. If specifying the output format, only one flag may
		be used. Using multiple output format flags will result in an error.
	`),
	Run: func(cmd *cobra.Command, args []string) {
		exiter := exit.FromCmd(cmd)

		// Error out if multiple output formats are specified.
		if flagJSON && flagYaml {
			exiter.Err("cannot use multiple formatting flags at once")
		}

		exiter.Err(pluginVersion(cmd.OutOrStdout()))
	},
}

func pluginVersion(out io.Writer) error {
	conn, client, err := utils.NewSynseGrpcClient(flagContext, flagTLSCert)
	if err != nil {
		return err
	}
	defer conn.Close()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	response, err := client.Version(ctx, &synse.Empty{})
	if err != nil {
		return err
	}

	printer := utils.NewPrinter(out, flagJSON, flagYaml, flagNoHeader)
	printer.SetIntermediateYaml()
	printer.SetHeader("VERSION", "SDK", "BUILD DATE", "OS", "ARCH")
	printer.SetRowFunc(pluginVersionRowFunc)

	return printer.Write(response)
}
