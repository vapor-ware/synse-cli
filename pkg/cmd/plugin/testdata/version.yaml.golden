pluginVersion: 3.2.1
sdkVersion: 3.0.0
